### PR TITLE
[#311] Remove voters limit and avoid returning tokens on flush 

### LINKED ISSUE #311: Remove voters limit and avoid returning tokens on flush
# Clarification and motivation

Currently as part of proposal `flush`ing the contract transfers (aka unstakes) back tokens used to vote on said proposal.
This is also the reason why there is a limit on the number of `voters` that can vote on any single proposal as, if they are too many, the call to `flush` would fail due to gas costs.

This behavior has been in the contract since basically its conception, and the logic itself relied on it in various places during its development, however currently that's no longer the case.
Moreover, this is actually problematic, because the cost of proposals handling increases linearly with the number of voters.

So, we want to to change this behavior to remove the voters limit and avoid returning tokens on `flush`.

# Acceptance criteria

1. on a `flush` call no token is returned to the `voters` (only to the proposer)
2. a new entrypoint exists that voters can call to get back the tokens staked on a proposal (after it was `flush`ed)
3. the limit on the number of voters per proposals is removed entirely
4. tests are updated accordingly

### PR DESCRIPTION
**Pull request by rinn7e**
## Description

Problem: The cost of proposals handling increases linearly with the
number of voters due to the fact that we unfreeze voters token on flush.

As a result, we want to to change this behavior to remove the voters
limit and avoid returning tokens on flush.

Solution: Remove returning voter tokens, remove voter limit, add
`Unfreeze_staked` entrypoints, and update the tests.

## Related issue(s)

Resolves #311

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: src/proposal.mligo [hunk 1]
```diff
@@ -88,6 +105,44 @@ let update_delegates (params, store : update_delegate_params * storage): return
...
+
+  // Remove the votes
+  let new_voter_proposal_map = Map.update proposal_key (None : (vote_status list) option) voter_proposal_map in
+
+  ( nil_op
+  , { store with
+      voters = Big_map.update Tezos.sender (Some new_voter_proposal_map) store.voters
+    }
+  )
+
```
Comment (rinn7e) [inline src/proposal.mligo:145]: @pasqu4le This PR is almost ready for review, only the tests are left to be done.

Before updating the tests, I want your advice on the current implementation. The new entrypoint is call `Unfreeze_unstaked`. Currently it only targets one proposal at a time, my questions are:
1. Do we plan to allow the user to unstake all their votes (on flushed proposals) all at once or not?
2. If the voter that call does not actually vote on that proposal, do we want to fail the call or let it through (by unfreeze nothing)? 
3. Do we keep their votes as history after `Unfreeze_unstaked` is called or removed them?

Hope to get your advice on this.

Comment (rinn7e) [reply src/proposal.mligo]: @pasqu4le I agree with all your points. I have push the final changes as well as updating the tests now, this should be ready for review. The only thing left is updating the docs, which will be what I'm working on next.

The new entrypoint is call `Unfreeze_staked`, let me know if the name is good enough as well.

Comment (pasqu4le) [reply src/proposal.mligo]: @rinn7e what do you think of `unfreeze_vote` as a name instead, given that it can only be used for tokens that were frozen for voting?

**Comment by sras** [comment]
@rinn7e Since we still store voter_map in the proposal structure, doesn't cost of decoding a single proposal still increase linearly with the number of voters who have voted in that proposal?

**Comment by rinn7e** [comment]
> @rinn7e Since we still store voter_map in the proposal structure, doesn't cost of decoding a single proposal still increase linearly with the number of voters who have voted in that proposal?

That's true, I'm thinking of storing the voters in storage as a big_map as well, but decide not to do it. If it is indeed a better choice to move the voter out of the proposal, I'll go with that approach.

File: src/types.mligo [hunk 2]
```diff
@@ -166,6 +160,7 @@ type storage =
   ; extra : contract_extra
   ; proposals : (proposal_key, proposal) big_map
   ; proposal_key_list_sort_by_level : (blocks * proposal_key) set
+  ; voters : (address * proposal_key, vote_status) big_map
```
Comment (rinn7e) [inline src/types.mligo:163]: @sras I've made the changes according to your feedback, this should be good now.

File: src/types.mligo [hunk 3]
```diff
@@ -223,6 +234,7 @@ type forbid_xtz_params =
   | Freeze of freeze_param
   | Unfreeze of unfreeze_param
   | Update_delegate of update_delegate_params
+  | Unfreeze_vote of unfreeze_vote_param
```
Comment (sras) [inline src/types.mligo:237]: Why is this called `Unfreeze` when what it is doing seems to be just unstaking staked frozen tokens back to unstaked frozen tokens. Is any 'Unfreez' ing happening here?

File: src/types.mligo [hunk 4]
```diff
@@ -51,8 +51,16 @@ type nonce = nat
...
 
-// Stores all voter data for a proposal
-type voter_map = (address * vote_type, nat) map
+type vote_item =
+  { vote_amount : nat
+  ; vote_type : vote_type
+  }
+
+type vote_status =
+  { votes : vote_item list
```
Comment (sras) [inline src/types.mligo:60]: I see that we have replced `voter_map` with `vote_item list`, I suppose this is done to track the total number of staked tokens. If that is the case can't we just add a `nat` field to `vote_status` to keep track of staked tokens, and increment it during each `vote` operation for this user, and keep the `votes` field as `voter_map` itself?

File: src/types.mligo [hunk 5]
```diff
@@ -94,10 +102,11 @@ type proposal =
   // ^ address of the proposer
   ; proposer_frozen_token : nat
   // ^ amount of frozen tokens used by the proposer, exluding the fixed fee
-  ; voters : voter_map
-  // ^ voter data
   ; quorum_threshold: quorum_threshold
   // ^ quorum threshold at the cycle in which proposal was raised
+  ; is_proposer_unfreezed: bool
```
Comment (sras) [inline src/types.mligo:107]: If this indicate that the proposal is either flushed or dropped, then I think it should be named accordingly.

File: src/types.mligo (hunk 3 above)
Comment (rinn7e) [reply src/types.mligo:237]: That's true, the term `unfreeze` on `flush` was used a lot previously, but the correct term should be `unstake` on `flush`, since we have a separate `unfreeze` entrypoint now. 

I'll rename this to `unstake` instead.

File: src/types.mligo (hunk 4 above)
Comment (rinn7e) [reply src/types.mligo:60]: `type voter_map = (address * vote_type, nat) map` is replaced by `(address * proposal_key, vote_status) big_map`, and is moved outside of the `proposal` to `storage`.

`vote_status` type is as followed:
```
type vote_status =
  { votes : vote_item list
  ; is_unfreezed : bool
  }

type vote_item =
  { vote_amount : nat
  ; vote_type : vote_type
  }

 ```
 
 I'm not sure exactly what your suggestion is, but I guess you mean that it is better to just use `nat` instead of a `list` there.
 
 ```
 type vote_status =
  { upvotes : nat
  ; downvotes: nat
  ; is_unfreezed : bool
  }
```

I could do it this way as well, but I think storing each upvote and downvote as a list could be useful in the future but I may be wrong.

File: src/types.mligo (hunk 3 above)
Comment (rinn7e) [reply src/types.mligo]: I've pushed the changes, have to squash the commits due to fixing conflicts when rebasing.

File: src/types.mligo (hunk 4 above)
Comment (sras) [reply src/types.mligo:60]:
```
type vote_status =
 { upvotes : nat
 ; downvotes: nat
 ; is_unfreezed : bool
 }
```
Ah, yes, that would work better.

>I think storing each upvote and downvote as a list could be useful in the future but I may be wrong.

But we didn't have that capability as of now, right?

Comment (rinn7e) [reply src/types.mligo:60]: > But we didn't have that capability as of now, right?
The current implementation already stores upvote and downvote as a list already. You mean we don't have the usage of it now right? 

@pasqu4le Since you suggest keeping the vote history, what do you think about @sras feedback? With the new feedback, we still keep track of the vote history but we won't know how many times the voter upvote/downvote a proposal (which may or may not be useful).

Comment (sras) [reply src/types.mligo:60]: >The current implementation already stores upvote and downvote as a list already.

@rinn7e Oh, by current implementation, I meant the version in master branch. AFAIU I that version, we don't keep the votes in a list.

Comment (rinn7e) [reply src/types.mligo:60]: I mean the implementation in this PR already store them in a list.

This is due to the suggestion 3 in here https://github.com/tezos-commons/baseDAO/pull/313#discussion_r767946169

File: docs/specification.md [hunk 6]
```diff
@@ -681,6 +677,27 @@ Parameter (in Michelson):
...
+
+Unstake_vote of unstake_vote_param
+```
+
+Parameter (in Michelson):
+```
+(list %unstake_vote bytes)
+```
+
+- Unfreeze voter's tokens for proposals that are already flushed.
```
Comment (sras) [inline docs/specification.md:695]: >Unfreeze voter's tokens

Probably we should also say "Unstake" here as well..

File: haskell/test/Test/Ligo/BaseDAO/Proposal/Propose.hs [hunk 7]
```diff
@@ -532,3 +536,66 @@ proposalBoundedValue originateFn = do
...
+  withSender dodOwner2 $ call dodDao (Call @"Unstake_vote") [key1]
+    & expectFailedWith voterAlreadyUnstake
+
+  -- Proposal key2 is not yet flush.
+  withSender dodOwner2 $ call dodDao (Call @"Unstake_vote") [key2]
+    & expectFailedWith proposalIsNotFlushed
+
+  -- owner 1 does not vote on the proposal
+  withSender dodOwner1 $ call dodDao (Call @"Unstake_vote") [key1]
+    & expectFailedWith voterDoesNotExist
```
Comment (sras) [inline haskell/test/Test/Ligo/BaseDAO/Proposal/Propose.hs:601]: AFAIU this test does not actually test if the "Unstaking" worked. Right?

File: src/types.mligo (hunk 4 above)
Comment (sras) [reply src/types.mligo:60]: Oh. I didn't see that comment. Hope he can provide some context.

File: docs/specification.md (hunk 6 above)
Comment (rinn7e) [reply docs/specification.md]: Nice catch, updated this.

File: haskell/test/Test/Ligo/BaseDAO/Proposal/Propose.hs (hunk 7 above)
Comment (rinn7e) [reply haskell/test/Test/Ligo/BaseDAO/Proposal/Propose.hs:619]: I added `freeze_history` comparison now, this should test if unstaking works properly.

Comment (sras) [reply haskell/test/Test/Ligo/BaseDAO/Proposal/Propose.hs:619]: Can we also add a test to check that the unstaking works when the proposal is dropped?

Comment (rinn7e) [reply haskell/test/Test/Ligo/BaseDAO/Proposal/Propose.hs:619]: Make sense, I've added a case for when the proposal is dropped as well.

File: src/proposal.mligo [hunk 8]
```diff
@@ -231,14 +272,16 @@ let unfreeze_proposer_and_voter_token
...
+  // Do the unstake for the proposer
   let store = unstake_tk(tokens, burn_amount, proposal.proposer, period, store) in
 
-  // unfreeze_voter_token
-  let do_unfreeze = fun
-        ( store, voter_with_vote : storage * ((address * vote_type) * nat))
-          -> unstake_tk(voter_with_vote.1, 0n, voter_with_vote.0.0, period, store) in
+  // Update `is_flushed_or_dropped`
+  let new_proposals = Map.add proposal_key
+        ({proposal with is_flushed_or_dropped = true}) store.proposals in
```
Comment (pasqu4le) [inline src/proposal.mligo:281]: Isn't it odd that this happens in a function called `unstake_proposer_token`?

I mean, by the name they don't seem related

File: src/types.mligo (hunk 4 above)
Comment (pasqu4le) [reply src/types.mligo:60]: > Hope he can provide some context

I mentioned this elsewhere as well, but this was a requested feature: #151

Comment (sras) [reply src/types.mligo:60]: @pasqu4le the issue is that version in master branch stores the information in  a map, which in this PR it is changed to a list. So basically the difference is that earlier we indeed stored who voted up or down, but didn't store how many times they voted. That is if they voted two times with 5 upvotes, then we only stored that they voted with 10 upvotes, but not that it was 2 times, with 5 votes each. But this PR adds this additional information, because the votes are stored in a list, so each instance of a vote is stored, which is more information than we need.

This is the change that I am pointing out as being new.

Comment (rinn7e) [reply src/types.mligo:60]: Regarding this point, I'm good with both combining the votes (like the previous implementation) or keeping it as a list (this PR implementation). I use the list simply because it could be useful, but if it isn't and cost too much, I'll go with combining them instead.

File: src/types.mligo [hunk 9]
```diff
@@ -94,10 +102,11 @@ type proposal =
   // ^ address of the proposer
   ; proposer_frozen_token : nat
   // ^ amount of frozen tokens used by the proposer, exluding the fixed fee
-  ; voters : voter_map
-  // ^ voter data
   ; quorum_threshold: quorum_threshold
   // ^ quorum threshold at the cycle in which proposal was raised
+  ; is_flushed_or_dropped: bool
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply src/types.mligo:107]: At first, I didn't think there is a way to calculate it, but now that you mention it, I think it is possible by checking `proposal_key_list_sort_by_level`. I'll make the change on this.

File: src/base_DAO.mligo [hunk 10]
```diff
@@ -25,6 +25,7 @@ let requiring_no_xtz (param, store, config : forbid_xtz_params * storage * confi
     | Freeze p               -> freeze(p, config, store)
     | Unfreeze p             -> unfreeze(p, config, store)
     | Update_delegate p      -> update_delegates(p, store)
+    | Unstake_vote p        -> unstake_vote(p, config, store)
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply src/base_DAO.mligo]: Fixed.

File: src/types.mligo [hunk 11]
```diff
@@ -94,10 +102,11 @@ type proposal =
...
+  // ^ If true, it indicates that the proposal is either flushed or dropped.
+  // Allow voters to call `Unstake_vote` on this proposal.
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply src/types.mligo]: Updated the config section.

File: scripts/generate_error_code.hs [hunk 12]
```diff
@@ -99,6 +98,12 @@ invalidInputErrors = errorsEnumerate 100
     :? ("Executing the proposal's decision lambda results in failure.", NoArg)
   , "entrypoint_not_found"
     :? ("The chosen custom entrypoint does not exist.", NoArg)
+  , "proposal_is_not_flushed"
+    :? ("The voter is trying to call `Unstake_vote` on proposals which are not flushed yet.", NoArg)
+  , "voter_does_not_exist"
+    :? ("The voter calling `Unstake_vote` does not vote on the proposal.", NoArg)
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply scripts/generate_error_code.hs]: Updated.

File: scripts/generate_error_code.hs [hunk 13]
```diff
@@ -99,6 +98,12 @@ invalidInputErrors = errorsEnumerate 100
...
+  , "voter_already_unstake"
+    :? ("The voter is trying to call `Unstake_vote` on proposals which his/her tokens have already been unstaked.", NoArg)
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply scripts/generate_error_code.hs]: Updated.

File: scripts/generate_error_code.hs [hunk 14]
```diff
@@ -99,6 +98,12 @@ invalidInputErrors = errorsEnumerate 100
     :? ("Executing the proposal's decision lambda results in failure.", NoArg)
   , "entrypoint_not_found"
     :? ("The chosen custom entrypoint does not exist.", NoArg)
+  , "proposal_is_not_flushed"
+    :? ("The voter is trying to call `Unstake_vote` on proposals which are not flushed yet.", NoArg)
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply scripts/generate_error_code.hs]: I've renamed this to `unstake_invalid_proposal` to make it clearer. Let me know if this is good enough.

File: docs/specification.md [hunk 15]
```diff
@@ -94,9 +94,6 @@ type config =
```
[the remark this answers is not in the document]
Comment (rinn7e) [reply docs/specification.md:94]: Updating the voting section part as well as the flush description. I'm not sure where else to update.

File: src/types.mligo (hunk 9 above)
[the remark this answers is not in the document]
Comment (rinn7e) [reply src/types.mligo]: I've removed `is_flushed_or_dropped` now and calculate it by checking `proposal_key_list_sort_by_level` instead.

File: src/proposal.mligo (hunk 8 above)
Comment (rinn7e) [reply src/proposal.mligo]: Since I've removed `is_flushed_or_dropped`, I also this part as well.

File: src/types.mligo (hunk 4 above)
Comment (pasqu4le) [reply src/types.mligo:60]: Ah I see, I didn't understand what was the difference, sorry.

AFAIK we only _need_ to keep the total amount of upvotes/downvotes and not the historic list of votes submitted.

I agree that we shouldn't keep this as we have no use for it now, so it looks like indeed:
```
type vote_status =
 { upvotes : nat
 ; downvotes: nat
 ; is_unfreezed : bool
 }
```
would be more efficient and work out just fine for this.

File: docs/specification.md [hunk 16]
```diff
@@ -681,6 +673,27 @@ Parameter (in Michelson):
...
+
+Unstake_vote of unstake_vote_param
+```
+
+Parameter (in Michelson):
+```
+(list %unstake_vote bytes)
+```
+
+- Unstake voter's tokens for proposals that are already flushed.
```
Comment (pasqu4le) [inline docs/specification.md:691]:
```suggestion
- Unstake voter's tokens for proposals that are already flushed or dropped.
```

File: docs/specification.md [hunk 17]
```diff
@@ -681,6 +673,27 @@ Parameter (in Michelson):
...
+- Fails with `unstake_invalid_proposal` error code if one of the proposals are not yet flushed or dropped.
+- Fails with `voter_does_not_exist` error code if the sender does not vote on the proposal.
```
Comment (pasqu4le) [inline docs/specification.md:693]:
```suggestion
- Fails with `voter_does_not_exist` error code if the sender did not vote on the proposal.
```

File: docs/specification.md [hunk 18]
```diff
@@ -681,6 +673,27 @@ Parameter (in Michelson):
...
+- Fails with `voter_unstaked` error code if the sender already called this entrypoint before.
```
Comment (pasqu4le) [inline docs/specification.md:694]:
```suggestion
- Fails with `voter_unstaked` error code if the sender already called this entrypoint before for one of the given proposals.
```

File: haskell/test/Test/Ligo/BaseDAO/Proposal.hs [hunk 19]
```diff
@@ -135,6 +131,9 @@ test_BaseDAO_Proposal =
 
       , testScenario "correctly track freeze history" $ scenario $
           checkFreezeHistoryTracking (originateLigoDaoWithConfigDesc dynRecUnsafe)
+
+      , testScenario "can unstake tokens after flush is called" $ scenario $
```
Comment (pasqu4le) [inline haskell/test/Test/Ligo/BaseDAO/Proposal.hs:135]: nitpick, but I suggest to change this to something like:
```suggestion
      , testScenario "tokens are unstaked correctly, only when possible" $ scenario $
```
simply because the scenario tests for incorrect cases and dropped proposals as well

File: src/proposal.mligo [hunk 20]
```diff
@@ -29,6 +29,13 @@ let check_if_proposal_exist (proposal_key, store : proposal_key * storage): prop
     then p
     else (failwith proposal_not_exist : proposal)
 
+[@inline]
+let check_if_proposal_flushed_or_dropped (proposal_key, store : proposal_key * storage): proposal =
+  let p = fetch_proposal (proposal_key, store) in
+  if Set.mem (p.start_level, proposal_key) store.proposal_key_list_sort_by_level
+    then (failwith unstake_invalid_proposal : proposal)
+    else p
```
Comment (pasqu4le) [inline src/proposal.mligo:37]: I would suggest to either:
- rename this function to `proposal_is_unstakeable`
- move its definition inside `unstake_vote_one` itself

for the reasons that this is only used by that function and that it throws `unstake_invalid_proposal` which is specific to this use.

File: src/proposal.mligo [hunk 21]
```diff
@@ -88,6 +112,42 @@ let update_delegates (params, store : update_delegate_params * storage): return
   , { store with delegates = List.fold update_delegate params store.delegates }
   )
 
+
+// Unstake voter's tokens on a proposal that is already been flushed. Fail if the voter does not vote
+// on that proposal, or voter has already unfreezed, or the proposal is not flushed.
```
Comment (pasqu4le) [inline src/proposal.mligo:117]:
```suggestion
// Unstake voter's tokens on a proposal that has already been flushed or dropped.
// Fail if the voter did not vote on that proposal, or voter has already unfreezed, or
// the proposal was not yet flushed nor dropped.
```

File: docs/specification.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (pasqu4le) [reply docs/specification.md:94]: @rinn7e I checked the spec and I don't see any other change necessary :ok_hand:

File: docs/specification.md (hunk 16 above)
Comment (rinn7e) [reply docs/specification.md]: Updated.

File: docs/specification.md (hunk 17 above)
Comment (rinn7e) [reply docs/specification.md]: Updated.

File: docs/specification.md (hunk 18 above)
Comment (rinn7e) [reply docs/specification.md]: Updated.

File: haskell/test/Test/Ligo/BaseDAO/Proposal.hs (hunk 19 above)
Comment (rinn7e) [reply haskell/test/Test/Ligo/BaseDAO/Proposal.hs]: Make sense, I've updated this.

File: src/proposal.mligo (hunk 20 above)
Comment (rinn7e) [reply src/proposal.mligo]: I've moved this to `unstake_vote_one` itself now.

File: src/proposal.mligo (hunk 21 above)
Comment (rinn7e) [reply src/proposal.mligo]: Updated.

File: src/types.mligo (hunk 4 above)
Comment (rinn7e) [reply src/types.mligo]: I've updated this. I think this is ready for re-review.

File: src/types.mligo (hunk 2 above)
Comment (pasqu4le) [reply src/types.mligo:156]: Minor point, but wouldn't this field be better called `votes`, since it track all kinds of vote info (not just who voted)?

File: src/proposal.mligo [hunk 22]
```diff
@@ -58,6 +58,23 @@ let ensure_proposal_is_unique (propose_params, store : propose_params * storage)
...
+        let new_freeze_history = Big_map.update addr (Some(fh)) store.freeze_history in
+        let new_total_supply =
+          match Michelson.is_nat (store.frozen_total_supply - burn_amount) with
+            Some new_total_supply -> new_total_supply
+          | None -> (failwith("BAD_STATE") : nat) in
+        { store with
+            freeze_history = new_freeze_history
+          ; frozen_total_supply = new_total_supply
+        }
+    | None -> (failwith("BAD_STATE") : storage)
```
Comment (pasqu4le) [inline src/proposal.mligo:76]: Shouldn't this be using the error code?
```suggestion
          | None -> (failwith bad_state : nat) in
        { store with
            freeze_history = new_freeze_history
          ; frozen_total_supply = new_total_supply
        }
    | None -> (failwith bad_state : storage)
```

File: src/proposal.mligo [hunk 23]
```diff
@@ -88,6 +105,43 @@ let update_delegates (params, store : update_delegate_params * storage): return
...
+            then (failwith unstake_invalid_proposal : unit)
+            else unit in
+
+  // Check if voter exist.
+  let vote_status = match Big_map.find_opt (Tezos.sender, proposal_key) store.voters with
+      | Some v -> v
+      | None -> (failwith voter_does_not_exist : vote_status) in
+
+  // Ensure that the voter has not called `unstake_vote` before.
+  let _ =  if vote_status.is_unstaked = true
```
Comment (pasqu4le) [inline src/proposal.mligo:126]: Not sure if ligo optimizes it away, but this could be:
```suggestion
  let _ =  if vote_status.is_unstaked
```

**Review by pasqu4le — APPROVED** [review]
Minor comments but otherwise LGTM

Don't forget to clean up the commit history tho!

File: src/proposal.mligo [hunk 24]
```diff
@@ -88,6 +105,43 @@ let update_delegates (params, store : update_delegate_params * storage): return
...
+      then (failwith voter_unstaked : unit)
+      else unit in
+
+  // Do the unstake
+  let tokens = vote_status.upvotes + vote_status.downvotes in
+  let store = unstake_tk(tokens, 0n, Tezos.sender, config.period, store) in
+
+  // Mark `is_unstaked` on voter's vote.
+  { store with
+      voters = Big_map.add (Tezos.sender, proposal_key) ({vote_status with is_unstaked = true}) store.voters
```
Comment (sras) [inline src/proposal.mligo:136]: I wonder if there is still a reason to maintain the voter in voters list after the voter has unstaked their token, @pasqu4le ?

File: docs/specification.md [hunk 25]
```diff
@@ -304,10 +299,7 @@ Moreover the proposal to vote on must have been submitted in the proposing `stag
 immediately preceding and the voter must have frozen his tokens in one of the
 preceding `stage`s.
 
-Each vote stakes one frozen token. Staked tokens cannot be unfreezed till they
-are unstaked when the associated proposal is flushed. The number of staked tokens
-only depend on the number of votes, and does not depend on whether the vote is in
-favor or against a proposal.
+Each vote stakes one frozen token. Staked tokens cannot be unfreezed till the associated proposal is flushed and the tokens are unstaked by calling `unstake_vote`. The number of staked tokens only depend on the number of votes, and does not depend on whether the vote is in favor or against a proposal.
```
Comment (sras) [inline docs/specification.md:302]: > associated proposal is flushed..

or dropped, right?

File: docs/specification.md [hunk 26]
```diff
@@ -610,6 +601,7 @@ Parameter (in Michelson):
   - If proposal got accepted:
     - The return amount for the proposer is equal to or less than the sum of the proposer frozen tokens and the fee paid for the proposal.
     - The return amount for each voters is equal to or less than the voter's frozen tokens.
+  - The token return to voters are not immediate. The voters should call `unstake_vote` to get their tokens after `flush` is called.
```
Comment (sras) [inline docs/specification.md:604]: We should also probably make it clear that voters should hold on to the proposal keys that they have voted on, if they want their staked tokens back.

File: src/proposal.mligo (hunk 24 above)
Comment (pasqu4le) [reply src/proposal.mligo:136]: good point, I suppose that we could remove the entries here (and the `is_unstaked` field entirely), but I [asked here](https://serokell.slack.com/archives/C02KY15P5GE/p1642015165003300?thread_ts=1637159479.009800&cid=C02KY15P5GE) to be sure.

@rinn7e if that's the case I guess my [previous comment](https://github.com/tezos-commons/baseDAO/pull/313#discussion_r783180286) would be different and I would instead consider calling this `storage` field `staked_votes` instead

File: docs/specification.md (hunk 26 above)
Comment (pasqu4le) [reply docs/specification.md:604]: They could also use an indexer/explorer for this tho.

For example, it would be sufficient to look at one's previous call to the `vote` entrypoint or traversing thie keys of the votes big_map.

In general I would avoid making suggestions in the specification when possible, given that this is supposed to be a document strictly describing what the contract does and how it behaves.

Comment (sras) [reply docs/specification.md:604]: Ok, I think we can at least say "should call `unstake_vote` with the proposal keys"

File: src/types.mligo (hunk 2 above)
Comment (rinn7e) [reply src/types.mligo]: I've updated this to `staked_votes` now.

File: src/proposal.mligo (hunk 22 above)
Comment (rinn7e) [reply src/proposal.mligo]: Updated.

File: src/proposal.mligo (hunk 24 above)
Comment (rinn7e) [reply src/proposal.mligo]: Seems like the response is fine with not keeping track of the voters. I've removed `is_unstaked` and rename to `staked_votes`. This should be good now.

Since we no longer know if the voter has called the entrypoint previously, I removed the `voter_unstaked` and mention its description as part of `voter_does_not exist` instead.

File: docs/specification.md (hunk 25 above)
Comment (rinn7e) [reply docs/specification.md]: Updated.

File: docs/specification.md (hunk 26 above)
Comment (rinn7e) [reply docs/specification.md]: I've added `with the proposal keys` part.

File: src/types.mligo [hunk 27]
```diff
@@ -51,10 +51,9 @@ type nonce = nat
 // Represents whether a voter has voted against (false) or for (true) a given proposal.
 type vote_type = bool
 
-type vote_status =
+type staked_vote =
  { upvotes : nat
  ; downvotes: nat
```
Comment (sras) [inline src/types.mligo:56]: Do we need the separate `upvotes`, `downvotes` fields now?

Comment (rinn7e) [reply src/types.mligo]: I guess we don't need to separate them anymore, but it may still be useful for the test (Easier to do the check).

I combine them anyway, you can check it now.

Comment (sras) [reply src/types.mligo]: LGTM now. But maybe @pasqu4le also better signoff this change as well.

**Review by sras — APPROVED** [review]

File: src/types.mligo (hunk 27 above)
Comment (pasqu4le) [reply src/types.mligo]: Yep, makes sense to me too.

This also was only used to keep historical data and is better off removed now :ok_hand:

File: src/proposal.mligo [hunk 28]
```diff
@@ -88,6 +105,37 @@ let update_delegates (params, store : update_delegate_params * storage): return
...
+
+  // Check if voter exist.
+  let staked_vote_amount = match Big_map.find_opt (Tezos.sender, proposal_key) store.staked_votes with
+      | Some v -> v
+      | None -> (failwith voter_does_not_exist : staked_vote) in
+
+  // Do the unstake
+  let store = unstake_tk(staked_vote_amount, 0n, Tezos.sender, config.period, store) in
+
+  // Mark `is_unstaked` on voter's vote.
```
Comment (pasqu4le) [inline src/proposal.mligo:128]: nitpick:
```suggestion
  // Remove voter's vote from staked amounts
```

**Review by pasqu4le — APPROVED** [review]
